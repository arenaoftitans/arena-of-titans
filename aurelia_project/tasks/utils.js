import {CLIOptions} from 'aurelia-cli';


export function buildObjectWithKeys(keys) {
    return keys
        .map(key => ({ [key]: {} }))
        .reduce((acc, current) => Object.assign(acc, current), {});
}


export function dumpAsExportedData(data) {
    return 'export default ' + JSON.stringify(data, null, 4) + ';\n';
}


export function flattenArray(array) {
    return array.reduce((acc, current) => acc.concat(current), []);
}


export function getVersion() {
    return CLIOptions.getFlagValue('version') || 'latest';
}


function getObjectAtPath(obj, path) {
    for (let pathElement of path) {
        if (!(pathElement in obj)) {
            obj[pathElement] = {};
        }
        obj = obj[pathElement];
    }

    return obj;
}


export function getRollbar() {
    const configPath = CLIOptions.getFlagValue('rollbar');
    const env = CLIOptions.getEnvironment();

    if (!configPath) {
        console.warn('Rollbar configuration was not supplied.');
        return {};
    }

    const rollbar = require(configPath);
    const envConfiguration = rollbar[env];
    if (!envConfiguration) {
        console.warn(`Rollbar is not available for ${env}`);
        return {};
    }

    return envConfiguration.front;
}


export function getTemplatesVariables() {
    return {
        env: CLIOptions.getEnvironment(),
        rollbar: getRollbar(),
        version: getVersion(),
    };
}


export function insertInto(obj, path, value) {
    const insertPath = path.split('.');
    const insertKey = insertPath.pop();
    obj = getObjectAtPath(obj, insertPath);
    obj[insertKey] = value;
}
