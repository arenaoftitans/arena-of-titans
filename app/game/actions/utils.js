/*
 * Copyright (C) 2015-2020 by Last Run Contributors.
 *
 * This file is part of Arena of Titans.
 *
 * Arena of Titans is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Arena of Titans is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with Arena of Titans. If not, see <http://www.gnu.org/licenses/>.
 *
 */

import { Container } from "aurelia-framework";
import { Api } from "../services/api";
import { BOARD_MOVE_MODE } from "../constants";
import { Popup } from "../services/popup";

const toCamel = s => {
    return s.replace(/([-_][a-z])/gi, $1 => {
        return $1
            .toUpperCase()
            .replace("-", "")
            .replace("_", "");
    });
};

const isObject = function(o) {
    return o === Object(o) && !Array.isArray(o) && typeof o !== "function";
};

export const keysToCamel = function(o) {
    if (isObject(o)) {
        const n = {};

        Object.keys(o).forEach(k => {
            n[toCamel(k)] = keysToCamel(o[k]);
        });

        return n;
    } else if (Array.isArray(o)) {
        return o.map(i => {
            return keysToCamel(i);
        });
    }

    return o;
};

export function getApi() {
    return Container.instance.get(Api);
}

export function displayPopup(popupType, options) {
    return Container.instance.get(Popup).display(popupType, options);
}

export function getEmptyCurrentTurn() {
    return {
        selectedCard: null,
        selectedTrump: null,
        boardMode: BOARD_MOVE_MODE,
        possibleSquares: [],
        mustAutoAskPossibleSquares: false,
        trumpInfosData: {
            infos: {},
            type: null,
        },
    };
}
