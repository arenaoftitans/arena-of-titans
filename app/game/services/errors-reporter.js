/*
* Copyright (C) 2015-2018 by Arena of Titans Contributors.
*
* This file is part of Arena of Titans.
*
* Arena of Titans is free software: you can redistribute it and/or modify
* it under the terms of the GNU Affero General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* Arena of Titans is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Affero General Public License for more details.
*
* You should have received a copy of the GNU Affero General Public License
* along with Arena of Titans. If not, see <http://www.gnu.org/licenses/>.
*/

import { inject } from 'aurelia-framework';
import { EventAggregatorSubscriptions } from './utils';
import { Popup } from '../widgets/popups/popup';


@inject(EventAggregatorSubscriptions, Popup)
export class ErrorsReporter {
    constructor(eas, popup) {
        this._eas = eas;
        this._popup = popup;
    }

    enable() {
        // We disable just to be sure there are no listeners registered.
        // If we don't, we may register them twice.
        this.disable();

        this._eas.subscribe('aot:api:error', data => {
            let popupData = {
                isFatal: data.isFatal,
                translate: {
                    messages: {
                        message: data.message,
                    },
                },
            };
            this._popup.display('error', popupData).then(() => {
                if (/\/game\/.+\/create\/.+/.test(location.pathname)) {
                    location.reload();
                }
            });
        });
    }

    disable() {
        this._eas.dispose();
    }
}
