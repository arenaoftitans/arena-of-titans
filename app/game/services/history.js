/*
* Copyright (C) 2016 by Arena of Titans Contributors.
*
* This file is part of Arena of Titans.
*
* Arena of Titans is free software: you can redistribute it and/or modify
* it under the terms of the GNU Affero General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* Arena of Titans is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Affero General Public License for more details.
*
* You should have received a copy of the GNU Affero General Public License
* along with Arena of Titans. If not, see <http://www.gnu.org/licenses/>.
*/

import { inject, NewInstance } from 'aurelia-framework';
import { EventAggregatorSubscriptions } from './utils';
import { Api } from './api';


@inject(Api, NewInstance.of(EventAggregatorSubscriptions))
export class History {
    _api;
    _ea;

    constructor(api, eas) {
        this._api = api;
        this._eas = eas;
        this.init();
    }

    init() {
        this._eas.dispose();
        this._eas.subscribe('aot:api:player_played', message => {
            this._addEntry(message.last_action);
        });
        // Map each players to his/her two last played cards.
        this._history = {};

        return this._api.onReconnectDefered.then(message => {
            let history = message.history;
            if (history) {
                for (let playerHistory of history) {
                    if (playerHistory) {
                        for (let action of playerHistory) {
                            this._addEntry(action);
                        }
                    }
                }
            }
        });
    }

    _addEntry(action) {
        let lastAction = action || {};
        let playerIndex = lastAction.player_index;
        let hist = this._getHistoryForPlayer(playerIndex);

        if (lastAction.card) {
            if (hist.length >= 2) {
                // Remove oldest element
                hist.shift();
            }
            hist.push(lastAction.card);
        }
    }

    _getHistoryForPlayer(index) {
        if (!this._history.hasOwnProperty(index)) {
            this._history[index] = [];
        }

        return this._history[index];
    }

    getLastPlayedCards(index) {
        return this._history[index] || [];
    }
}
