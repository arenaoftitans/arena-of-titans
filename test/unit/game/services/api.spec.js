/*
* Copyright (C) 2015-2016 by Arena of Titans Contributors.
*
* This file is part of Arena of Titans.
*
* Arena of Titans is free software: you can redistribute it and/or modify
* it under the terms of the GNU Affero General Public License as published by
* the Free Software Foundation, either version 3 of the License, or
* (at your option) any later version.
*
* Arena of Titans is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
* GNU Affero General Public License for more details.
*
* You should have received a copy of the GNU Affero General Public License
* along with Arena of Titans. If not, see <http://www.GNU Affero.org/licenses/>.
*/

import '../../setup';
import { Api } from '../../../../app/game/services/api';
import { NotifyStub, StorageStub, WsStub } from '../../utils';


describe('services/api', () => {
    let mockedStorage;
    let mockedWs;
    let mockedConfig;
    let mockedNotify;
    let sut;
    let rt;

    beforeEach(() => {
        mockedStorage = new StorageStub();
        mockedWs = new WsStub();
        mockedConfig = {
            test: {
                debug: false
            }
        };
        mockedNotify = new NotifyStub();
        sut = new Api(mockedWs, mockedStorage, mockedConfig, mockedNotify);
        rt = sut.requestTypes;
    });

    it('should send game data to ws on initialize game', () => {
        let gameData = {
            player_name: 'Tester',
            hero: 'daemon',
            rt: rt.init_game
        };
        spyOn(mockedWs, 'send');

        sut.initializeGame('Tester', 'daemon');

        expect(mockedWs.send).toHaveBeenCalledWith(gameData);
    });

    it('should send slot data to ws', () => {
        let slot = {
            index: 0,
            player_name: '',
            state: 'OPEN'
        };
        spyOn(mockedWs, 'send');
        sut._game.slots = [];

        sut.addSlot();
        expect(mockedWs.send).toHaveBeenCalledWith({
            rt: rt.add_slot,
            slot: slot
        });
    });

    it('should register callbacks', () => {
        let cb = () => {
        };
        sut.on(rt.init_game, cb);

        expect(sut.callbacks[rt.init_game].length).toBe(1);
        expect(sut.callbacks[rt.init_game][0]).toBe(cb);
    });

    it('should deregister callbacks', () => {
        let index = sut.on(rt.init_game, () => {
        });
        sut.off(rt.init_game, index);

        expect(sut.callbacks[rt.init_game].length).toBe(1);
        expect(sut.callbacks[rt.init_game][0]).toBe(undefined);
    });

    it('should handle game initialized', () => {
        let gameInitializedMessage = {
            rt: sut.requestTypes.game_initialized,
            index: 0,
            is_game_master: true,
            player_id: 'player_id',
            game_id: 'game_id',
            slots: [{
                index: 0,
                name: 'Player 1',
                state: 'TAKEN'
            }]
        };
        let gameInitializedCallback = jasmine.createSpy('gameInitializedCallback');
        sut.on(rt.game_initialized, gameInitializedCallback);
        spyOn(sut, '_handleGameInitialized').and.callThrough();
        spyOn(mockedStorage, 'savePlayerId');

        sut._handleMessage(gameInitializedMessage);
        expect(sut._handleGameInitialized).toHaveBeenCalledWith(gameInitializedMessage);
        expect(mockedStorage.savePlayerId).toHaveBeenCalledWith(gameInitializedMessage.game_id, gameInitializedMessage.player_id);
        expect(sut.me.index).toBe(gameInitializedMessage.index);
        expect(sut.me.is_game_master).toBe(gameInitializedMessage.is_game_master);
        expect(sut.me.id).toBe(gameInitializedMessage.player_id);
        expect(sut.game.id).toBe(gameInitializedMessage.game_id);
        expect(sut.game.slots).toBe(gameInitializedMessage.slots);
        expect(gameInitializedCallback).toHaveBeenCalledWith(gameInitializedMessage);
    });

    it('should handle slot updated', () => {
        let slotUpdatedMessage = {
            rt: sut.requestTypes.slot_updated,
            slot: {
                index: 0,
                name: '',
                state: 'OPEN'
            }
        };
        let slotUpdatedCallback = jasmine.createSpy('slotUpdatedCallback');
        sut.on(rt.slot_updated, slotUpdatedCallback);
        spyOn(sut, '_handleSlotUpdated').and.callThrough();
        sut._game.slots = [];

        expect(sut._game.slots.length).toBe(0);
        sut._handleMessage(slotUpdatedMessage);
        expect(sut._handleSlotUpdated).toHaveBeenCalledWith(slotUpdatedMessage);
        expect(sut._game.slots.length).toBe(1);
        expect(sut._game.slots[0]).toBe(slotUpdatedMessage.slot);
        expect(sut._game.slots[0].name).toBe('');
        expect(slotUpdatedCallback).toHaveBeenCalledWith(slotUpdatedMessage);

        slotUpdatedMessage.slot.name = 'Player 1';
        slotUpdatedMessage.slot.state = 'TAKEN';
        sut._handleMessage(slotUpdatedMessage);
        expect(sut._game.slots.length).toBe(1);
        expect(sut._game.slots[0].name).toBe('Player 1');
        expect(sut._game.slots[0].state).toBe('TAKEN');
        expect(sut._game.slots[0].index).toBe(0);
        expect(slotUpdatedCallback).toHaveBeenCalledWith(slotUpdatedMessage);
    });

    it('should update name', () => {
        spyOn(mockedWs, 'send');

        sut._me = {
            name: 'Player 1',
            index: 0,
            hero: 'daemon',
        };
        sut._game = {
            slots: [{
                index: 0,
                player_name: 'Player 1',
                state: 'TAKEN',
            }]
        };
        sut.updateMe('New Name', 'reaper');

        expect(sut.me.hero).toBe('reaper');
        expect(sut.me.name).toBe('New Name');
        expect(sut.game.slots[0].hero).toBe('reaper');
        expect(sut.game.slots[0].player_name).toBe('New Name');
        expect(mockedWs.send).toHaveBeenCalledWith({
            rt: sut.requestTypes.slot_updated,
            slot: {
                hero: 'reaper',
                index: 0,
                player_name: 'New Name',
                state: 'TAKEN'
            }
        })
    });

    it('should send game data when joining game', () => {
        spyOn(mockedWs, 'send');

        sut.joinGame({gameId: 'the_game_id', name: 'Player 2', hero: 'daemon'});

        expect(mockedWs.send).toHaveBeenCalledWith({
            rt: sut.requestTypes.init_game,
            player_name: 'Player 2',
            game_id: 'the_game_id',
            player_id: undefined,
            hero: 'daemon',
        });
    });

    it('should send game data when joining game with a game id', () => {
        spyOn(mockedWs, 'send');

        sut.joinGame({gameId: 'the_game_id', playerId: 'player_id', hero: 'daemon'});

        expect(mockedWs.send).toHaveBeenCalledWith({
            rt: sut.requestTypes.init_game,
            player_name: undefined,
            game_id: 'the_game_id',
            player_id: 'player_id',
            hero: 'daemon',
        });
    });

    describe('joinGame', () => {
        it('should fetch the game id if neither name nor playerId is defined', () => {
            spyOn(mockedStorage, 'retrievePlayerId').and.returnValue('player_id');
            spyOn(mockedWs, 'send');

            sut.joinGame({gameId: 'game_id'});

            expect(mockedStorage.retrievePlayerId).toHaveBeenCalledWith('game_id');
            expect(mockedWs.send).toHaveBeenCalledWith({
                rt: sut.requestTypes.init_game,
                player_name: undefined,
                game_id: 'game_id',
                player_id: 'player_id',
                hero: undefined,
            });
        });

        it('should not fetch the if playerId is defined', () => {
            spyOn(mockedStorage, 'retrievePlayerId');

            sut.joinGame({gameId: 'game_id', playerId: 'player_id'});

            expect(mockedStorage.retrievePlayerId).not.toHaveBeenCalled();
        });

        it('should not fetch the if name is defined', () => {
            spyOn(mockedStorage, 'retrievePlayerId');

            sut.joinGame({gameId: 'game_id', name: 'Player 1'});

            expect(mockedStorage.retrievePlayerId).not.toHaveBeenCalled();
        });
    });

    it('should handle errors to display', () => {
        let message = {error_to_display: 'error'};
        let errorCb = jasmine.createSpy('errorCb');
        spyOn(sut, '_handleErrors').and.callThrough();
        spyOn(sut, '_callCallbacks');

        sut.onerror(errorCb);
        sut._handleMessage(message);

        expect(sut._handleErrors).toHaveBeenCalledWith(message);
        expect(sut._callCallbacks).not.toHaveBeenCalled();
        expect(errorCb).toHaveBeenCalledWith({message: message.error_to_display});
    });

    it('should handle errors to log', () => {
        let message = {error: 'error'};
        let errorCb = jasmine.createSpy('errorCb');
        spyOn(sut, '_handleErrors').and.callThrough();
        spyOn(sut, '_callCallbacks');
        spyOn(console, 'error');

        sut.onerror(errorCb);
        sut._handleMessage(message);

        expect(sut._handleErrors).toHaveBeenCalledWith(message);
        expect(sut._callCallbacks).not.toHaveBeenCalled();
        expect(errorCb).not.toHaveBeenCalled();
        expect(console.error).toHaveBeenCalledWith(message);
    });

    it('should create the game', () => {
        spyOn(mockedWs, 'send');

        sut._game = {
            slots: [
                {
                    index: 0,
                    player_name: 'Player 1',
                    hero: 'daemon',
                },
                {
                    index: 1,
                    player_name: 'Player 2',
                    hero: 'daemon',
                }
            ]
        };
        sut.createGame();

        expect(mockedWs.send).toHaveBeenCalledWith({
            rt: sut.requestTypes.create_game,
            create_game_request: [
                {
                    index: 0,
                    name: 'Player 1',
                    hero: 'daemon',
                },
                {
                    index: 1,
                    name: 'Player 2',
                    hero: 'daemon',
                }
            ]
        })
    });

    it('should update the game', () => {
        let elapsedTime = 12;
        let message = {
            your_turn: true,
            has_won: false,
            rank: -1,
            next_player: 0,
            active_trumps: [{
                player_index: 0,
                player_name: "Player 1",
                trumps: [{
                    name: 'Reinforcements'
                }]
            }, {
                player_index: 1,
                player_name: "Player 2",
                trumps: []
            }],
            hand: [{
                name: "King",
                color: "RED"
            }],
            elapsed_time: elapsedTime,
        };
        sut._me.index = 0;

        sut._updateGame(message);

        expect(sut._game.your_turn).toBe(true);
        expect(sut._game.next_player).toBe(0);
        expect(sut._game.active_trumps.length).toBe(2);
        expect(sut._game.active_trumps[0]).toBe(message.active_trumps[0]);
        expect(sut._game.active_trumps[1]).toBe(message.active_trumps[1]);
        expect(sut._me.hand).toEqual([
            {
                name: 'King',
                color: 'RED',
                img: 'sprite-movement-king-red'
            }
        ]);
        expect(sut._me.has_won).toBe(false);
        expect(sut._me.rank).toBe(-1);
        expect(sut._me.affecting_trumps).toEqual([{
            name: 'Reinforcements',
            img: '/assets/game/cards/trumps/reinforcements.png'
        }]);
        expect(sut._me.elapsed_time).toBe(elapsedTime);
    });

    it('should handle player played', () => {
        let message = {
            game_over: false,
            winners: [],
        };

        sut._handleGameOverMessage(message);

        expect(sut._game.game_over).toBe(false);
        expect(sut._game.winners.length).toBe(0);
    });

    it('should handle player played game over', () => {
        let message = {
            game_over: true,
            winners: ['Player 1', 'Player 2'],
        };
        spyOn(sut._gameOverDefered, 'resolve');

        sut._handleGameOverMessage(message);

        expect(sut._game.game_over).toBe(true);
        expect(sut._game.winners.length).toBe(2);
        expect(sut._gameOverDefered.resolve).toHaveBeenCalledWith(message.winners);
    });

    it('should handle game created data', () => {
        let message = {
            rt: sut.requestTypes.create_game,
            players: [{
                hero: 'daemon',
                index: 0,
                name: "Player 1"
            }, {
                hero: 'reaper',
                index: 1,
                name: "Player 2"
            }],
            trumps: [{
                name: "Red Tower",
                description: "Prevents a player to move on red squares.",
                cost: 0,
                duration: 1,
                must_target_player: true
            }]
        };
        spyOn(sut, '_updateGame');
        sut._handleMessage(message);

        expect(sut._updateGame).toHaveBeenCalledWith(message);
        expect(sut._game.players).toEqual({
            heroes: ['daemon', 'reaper'],
            indexes: [0, 1],
            names: ['Player 1', 'Player 2'],
            squares: [{}, {}]
        });
        expect(sut._me.trumps).toEqual([{
            name: "Red Tower",
            img: '/assets/game/cards/trumps/red-tower.png',
            description: "Prevents a player to move on red squares.",
            cost: 0,
            duration: 1,
            must_target_player: true
        }]);
    });

    it('should create game for debug', () => {
        spyOn(sut, 'initializeGame');
        spyOn(sut, 'addSlot').and.callThrough();
        spyOn(mockedWs, 'send');
        spyOn(sut, 'createGame');
        mockedConfig.test.debug = true;

        sut.createGameDebug();
        expect(sut.initializeGame).toHaveBeenCalledWith('Player 1');

        sut._handleMessage({
            rt: rt.game_initialized,
            game_id: 'game_id',
            player_id: 'player_id',
            index: 0,
            is_game_master: true,
            slots: [{
                player_name: 'Player 1'
            }]
        });
        expect(sut.addSlot).toHaveBeenCalled();

        sut._handleMessage({
            rt: rt.slot_updated,
            slot: {
                index: 1
            }
        });
        expect(sut.createGame).toHaveBeenCalled();
        expect(mockedWs.send).toHaveBeenCalledWith({
            rt: rt.add_slot,
            slot: {
                index: 1,
                player_name: 'Player 2',
                state: 'TAKEN'
            }
        });
    });

    describe('game', () => {
        it('should view possible movements', () => {
            spyOn(mockedWs, 'send');

            sut.viewPossibleMovements({name: 'King', color: 'red'});

            expect(mockedWs.send).toHaveBeenCalledWith({
                rt: sut.requestTypes.view,
                play_request: {
                    card_name: 'King',
                    card_color: 'red'
                }
            });
        });

        it('should play', () => {
            spyOn(mockedWs, 'send');

            sut.play({cardName: 'King', cardColor: 'red', x: '0', y: '0'});

            expect(mockedWs.send).toHaveBeenCalledWith({
                rt: sut.requestTypes.play,
                play_request: {
                    card_name: 'King',
                    card_color: 'red',
                    x: 0,
                    y: 0
                }
            });
        });

        it('should handle play and your turn', () => {
            let message = {
                rt: sut.requestTypes.play,
                your_turn: true
            };
            spyOn(sut, '_updateGame');
            spyOn(mockedNotify, 'clearNotifications');
            spyOn(mockedNotify, 'notifyYourTurn');
            sut._game.your_turn = true;

            sut._handleMessage(message);

            expect(sut._updateGame).toHaveBeenCalledWith(message);
            expect(mockedNotify.clearNotifications).not.toHaveBeenCalled();
            expect(mockedNotify.notifyYourTurn).toHaveBeenCalled();
        });

        it('should handle play and not your turn', () => {
            let message = {
                rt: sut.requestTypes.play,
                your_turn: false
            };
            spyOn(sut, '_updateGame');
            spyOn(mockedNotify, 'clearNotifications');
            spyOn(mockedNotify, 'notifyYourTurn');
            sut._game.your_turn = false;

            sut._handleMessage(message);

            expect(sut._updateGame).toHaveBeenCalledWith(message);
            expect(mockedNotify.clearNotifications).toHaveBeenCalled();
            expect(mockedNotify.notifyYourTurn).not.toHaveBeenCalled();
        });

        it('should reconnect', () => {
            let message = {
                rt: sut.requestTypes.play,
                reconnect: {
                    players: [
                        {
                            index: 0,
                            square: {x: 0, y: 0},
                            hero: 'daemon',
                        }
                    ],
                    trumps: [
                        {
                            name: 'Tower red'
                        }
                    ]
                }
            };
            spyOn(sut, '_updateGame');

            sut._handleMessage(message);

            expect(sut._updateGame).toHaveBeenCalled();
            expect(sut._game.players).toEqual({
                indexes: [0],
                names: [undefined],
                squares: [{x: 0, y: 0}],
                heroes: ['daemon'],
            });
            expect(sut._me.trumps).toEqual([
                {
                    name: 'Tower red',
                    img: '/assets/game/cards/trumps/tower-red.png'
                }
            ])
        });

        it('should ask name when reconnecting to a freed slot', () => {
            spyOn(sut._reconnectDefered, 'reject');

            let message = {
                rt: sut.requestTypes.game_initialized,
                index: -1,
            };

            sut._handleMessage(message);

            expect(sut._reconnectDefered.reject).toHaveBeenCalled();
        });

        it('should play trump without target', () => {
            spyOn(mockedWs, 'send');

            sut.playTrump({trumpName: 'Trump'});

            expect(mockedWs.send).toHaveBeenCalledWith({
                rt: sut.requestTypes.play_trump,
                play_request: {
                    name: 'Trump',
                    target_index: null
                }
            });
        });

        it('should play trump with a target', () => {
            spyOn(mockedWs, 'send');

            sut.playTrump({trumpName: 'Trump', targetIndex: 0});

            expect(mockedWs.send).toHaveBeenCalledWith({
                rt: sut.requestTypes.play_trump,
                play_request: {
                    name: 'Trump',
                    target_index: 0
                }
            });
        });

        it('should pass', () => {
            spyOn(mockedWs, 'send');

            sut.pass();

            expect(mockedWs.send).toHaveBeenCalledWith({
                rt: sut.requestTypes.play,
                play_request: {
                    pass: true
                }
            })
        });

        it('should discard', () => {
            spyOn(mockedWs, 'send');

            sut.discard({cardName: 'King', cardColor: 'red'});

            expect(mockedWs.send).toHaveBeenCalledWith({
                rt: sut.requestTypes.play,
                play_request: {
                    discard: true,
                    card_name: 'King',
                    card_color: 'red'
                }
            })
        });
    });
});
