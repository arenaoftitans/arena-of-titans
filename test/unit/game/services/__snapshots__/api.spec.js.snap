// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`api joinGame 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "game_id": "game-id",
        "hero": "Arline",
        "player_name": "Test player 2",
      },
      "rt": "JOIN_GAME",
    },
  ],
]
`;

exports[`api should create game 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "players": Array [],
      },
      "rt": "CREATE_GAME",
    },
  ],
]
`;

exports[`api should create lobby 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "hero": "Arline",
        "player_name": "Test player",
      },
      "rt": "CREATE_LOBBY",
    },
  ],
]
`;

exports[`api should discard car 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "card_color": "Blue",
        "card_name": "King",
        "discard": true,
      },
      "rt": "PLAY_CARD",
    },
  ],
]
`;

exports[`api should handle message 1`] = `
Object {
  "currentTurn": Object {
    "boardMode": "move_mode",
    "possibleSquares": Array [],
    "selectedCard": null,
    "selectedTrump": null,
    "trumpInfosData": Object {
      "infos": Object {},
      "type": null,
    },
  },
  "game": Object {
    "actions": Array [
      Object {
        "card": null,
        "description": "passed_turn",
        "initiator": Object {
          "index": 0,
          "name": "Edwin",
        },
        "specialAction": null,
        "target": null,
        "trump": null,
      },
    ],
    "board": Object {
      "updatedSquares": Array [],
    },
    "currentPlayerIndex": 1,
    "id": "fPuZBQDWQxGuKSo-YUY-hQ",
    "isOver": false,
    "nbTurns": 0,
    "players": Object {
      "0": Object {
        "activeTrumps": Array [],
        "hero": "Mirindrel",
        "index": 0,
        "isVisible": true,
        "name": "Edwin",
        "square": Object {
          "color": "RED",
          "x": 6,
          "y": 7,
        },
      },
      "1": Object {
        "activeTrumps": Array [],
        "hero": "Luni",
        "index": 1,
        "isVisible": true,
        "name": "AI 1",
        "square": Object {
          "color": "YELLOW",
          "x": 6,
          "y": 8,
        },
      },
    },
    "winners": Array [],
  },
  "me": Object {
    "activeTrumps": Array [],
    "availableTrumps": Array [
      Object {
        "applyOnInitiator": false,
        "color": null,
        "colors": Array [],
        "cost": 6,
        "deltaMoves": -1,
        "description": "Reduce the number of cards a player can play by 1.",
        "duration": 1,
        "img": "/dist/assets/game/cards/trumps/blizzard-4a67467324.png",
        "isPlayerVisible": true,
        "mustTargetPlayer": true,
        "name": "Blizzard",
        "preventTrumpsToModify": Array [],
        "requireSquareTarget": false,
      },
      Object {
        "applyOnInitiator": false,
        "color": null,
        "colors": Array [],
        "cost": 4,
        "deltaDuration": -1,
        "description": "Destroy towers and reduce duration for fortress",
        "duration": 1,
        "img": "/dist/assets/game/cards/trumps/ram-73a6490a4e.png",
        "isPlayerVisible": true,
        "mustTargetPlayer": false,
        "name": "Ram",
        "preventTrumpsToModify": Array [],
        "requireSquareTarget": false,
        "trumpNames": Array [
          "Fortress",
          "Tower",
        ],
      },
      Object {
        "applyOnInitiator": false,
        "color": "YELLOW",
        "colors": Array [],
        "cost": 7,
        "description": "Prevent the player to move on some colors.",
        "duration": 2,
        "img": "/dist/assets/game/cards/trumps/fortress-yellow-67a7f343b7.png",
        "isPlayerVisible": true,
        "mustTargetPlayer": true,
        "name": "Fortress",
        "preventTrumpsToModify": Array [],
        "requireSquareTarget": false,
      },
      Object {
        "applyOnInitiator": false,
        "color": null,
        "colors": Array [],
        "cost": 8,
        "deltaMoves": 1,
        "description": "Allow the player to play one more move.",
        "duration": 1,
        "img": "/dist/assets/game/cards/trumps/reinforcements-489d9fe1d5.png",
        "isPlayerVisible": true,
        "mustTargetPlayer": false,
        "name": "Reinforcements",
        "preventTrumpsToModify": Array [],
        "requireSquareTarget": false,
      },
    ],
    "canPowerBePlayed": false,
    "elapsedTime": 0,
    "gaugeValue": 0,
    "hand": Array [
      Object {
        "color": "RED",
        "description": "Move two squares in line or diagonal.",
        "img": "sprite-movement-assassin-red",
        "name": "Assassin",
      },
      Object {
        "color": "RED",
        "description": "Move one squares in line or diagonal. Can move on a square of any color.",
        "img": "sprite-movement-wizard-red",
        "name": "Wizard",
      },
      Object {
        "color": "BLUE",
        "description": "Move one square in L.",
        "img": "sprite-movement-knight-blue",
        "name": "Knight",
      },
      Object {
        "color": "BLUE",
        "description": "Move three squares in line.",
        "img": "sprite-movement-king-blue",
        "name": "King",
      },
      Object {
        "color": "YELLOW",
        "description": "Move one square in L.",
        "img": "sprite-movement-knight-yellow",
        "name": "Knight",
      },
    ],
    "hasRemainingMovesToPlay": true,
    "hasWon": false,
    "hero": "Mirindrel",
    "id": "xgjfvkDc3p1hGI+Dw7kRLA==",
    "index": 0,
    "name": "Edwin",
    "onLastLine": false,
    "power": Object {
      "cost": 8,
      "description": "Change the color of a square",
      "img": "/dist/assets/game/cards/powers/terraforming-f264420b50.png",
      "isPower": true,
      "mustTargetPlayer": false,
      "name": "Terraforming",
      "passive": false,
      "requireSquareTarget": true,
      "trumpCostDelta": 0,
    },
    "rank": -1,
    "specialAction": null,
    "trumpTargetIndexes": Array [
      0,
      1,
    ],
    "trumpsStatuses": Array [
      false,
      false,
      false,
      false,
    ],
    "yourTurn": false,
  },
}
`;

exports[`api should pass special action 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "auto": false,
        "cancel": true,
        "special_action_name": "Assassination",
      },
      "rt": "SPECIAL_ACTION_PLAY",
    },
  ],
]
`;

exports[`api should pass turn 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "auto": false,
        "pass": true,
      },
      "rt": "PLAY_CARD",
    },
  ],
]
`;

exports[`api should play card 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "card_color": "red",
        "card_name": "King",
        "x": 0,
        "y": 0,
      },
      "rt": "PLAY_CARD",
    },
  ],
]
`;

exports[`api should play special action 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "special_action_name": "Assassination",
        "target_index": 0,
        "x": 0,
        "y": 0,
      },
      "rt": "SPECIAL_ACTION_PLAY",
    },
  ],
]
`;

exports[`api should play trump 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "color": "black",
        "name": "Tower",
        "square": Object {},
        "target_index": 1,
      },
      "rt": "PLAY_TRUMP",
    },
  ],
]
`;

exports[`api should reconnect 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "game_id": Object {
          "gameId": "game-id",
          "hero": "Arline",
          "playerName": "Test player",
        },
        "player_id": undefined,
      },
      "rt": "RECONNECT",
    },
  ],
]
`;

exports[`api should update solt 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "slot": Object {
          "hero": undefined,
          "index": 1,
          "player_name": undefined,
          "state": "CLOSED",
        },
      },
      "rt": "UPDATE_SLOT",
    },
  ],
]
`;

exports[`api should view possible actions 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "special_action_name": "Assasination",
        "target_index": 1,
      },
      "rt": "SPECIAL_ACTION_VIEW_POSSIBLE_ACTIONS",
    },
  ],
]
`;

exports[`api should view possible movements 1`] = `
Array [
  Array [
    Object {
      "request": Object {
        "card_color": "blue",
        "card_name": "King",
      },
      "rt": "VIEW_POSSIBLE_SQUARES",
    },
  ],
]
`;
